compiler/rustc_mir_transform/src/dest_prop.rs

```diff
@@ -353,524 +336,564 @@ impl<'tcx> MutVisitor<'tcx> for Replacer<'tcx> {
+        }
+    }
+    // Deduplicate the reverse map
+    for (_, cands) in candidates_reverse.iter_mut() {
+        cands.sort();
+        cands.dedup();
+    }
+    Candidates { c: candidates, reverse: candidates_reverse }
 }
 
-struct FindAssignments<'a, 'tcx> {
+struct FindAssignments<'a, 'alloc, 'tcx> {
     body: &'a Body<'tcx>,
-    candidates: Vec<CandidateAssignment<'tcx>>,
-    ever_borrowed_locals: BitSet<Local>,
-    locals_used_as_array_index: BitSet<Local>,
+    candidates: &'alloc mut FxHashMap<Local, Vec<Local>>,
+    borrowed: &'a BitSet<Local>,
 }
 
-impl<'tcx> Visitor<'tcx> for FindAssignments<'_, 'tcx> {
-    fn visit_statement(&mut self, statement: &Statement<'tcx>, location: Location) {
+impl<'tcx> Visitor<'tcx> for FindAssignments<'_, '_, 'tcx> {
+    fn visit_statement(&mut self, statement: &Statement<'tcx>, _: Location) {
         if let StatementKind::Assign(box (
-            dest,
-            Rvalue::Use(Operand::Copy(src) | Operand::Move(src)),
+            lhs,
+            Rvalue::Use(Operand::Copy(rhs) | Operand::Move(rhs)),
         )) = &statement.kind
         {
-            // `dest` must not have pointer indirection.
-            if dest.is_indirect() {
+            if !is_constant(*lhs) || !is_constant(*rhs) {
                 return;
             }
 
-            // `src` must be a plain local.
-            if !src.projection.is_empty() {
+            // Because we always merge whole locals into places, there's no scenario in which these
+            // kinds of assignments could be removed.
+            if lhs.local == rhs.local {
                 return;
             }
 
-            // Since we want to replace `src` with `dest`, `src` must not be required.
-            if is_local_required(src.local, self.body) {
-                return;
-            }
-
-            // Can't optimize if either local ever has their address taken. This optimization does
-            // liveness analysis only based on assignments, and a local can be live even if its
-            // never assigned to again, because a reference to it might be live.
-            // FIXME: This can be smarter and take `StorageDead` into  account (which invalidates
-            // borrows).
-            if self.ever_borrowed_locals.contains(dest.local)
-                || self.ever_borrowed_locals.contains(src.local)
-            {
+            // As described at the top of the file, we do not go near things that have their address
+            // taken.
+            if self.borrowed.contains(lhs.local) || self.borrowed.contains(rhs.local) {
                 return;
             }
 
-            assert_ne!(dest.local, src.local, "self-assignments are UB");
-
-            // We can't replace locals occurring in `PlaceElem::Index` for now.
-            if self.locals_used_as_array_index.contains(src.local) {
+            let Some((mut src, mut dest)) = get_bases(*lhs, *rhs) else {
```

Comment:
The `get_bases` allows identical projections on both sides. Is that case reachable?

Could you please add test or if not possible restrict this further to reject projections altogether.

Reply:
Yeah, this is reachable by something like:

```rust
x = (0, 5);
y.0 = x.0;
y.1 = 10;
dbg!(y);
```

I'll add this test when I go through and improve tests in general. I'll keep this open to track that

Reply:
I would suspect not because `a.x = b.y;` would go through an intermediate temporary `_tmp = b.y; a.x = _tmp`.

Reply:
Oh, good point - this would have worked and been necessary before we removed projecting into non-full locals, but now you might be right. I'll test this out, and if it doesn't actually fire remove it

Reply:
I tested, and you're indeed right. I've adjusted the check